src/core/middlewares/regquota/handler.go

```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to check Manifest existence %v", err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.Exist = exist
+			if exist {
+				if af.Digest != rqh.mfInfo.Digest {
+					rqh.mfInfo.DigestChanged = true
 				}
-				mfDigest = desc.Digest.String()
-				mfSize = desc.Size
-				log.Infof("manifest digest... %s", mfDigest)
-				log.Infof("manifest size... %v", mfSize)
 			}
+			quotaRes, err := rqh.tryRequireQuota()
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Errorf("Cannot get quota for the manifest %v", err)
+				http.Error(rw, util.MarshalError("StatusNotAcceptable", fmt.Sprintf("Cannot get quota for the manifest %v", err)), http.StatusNotAcceptable)
+				return
+			}
+			rqh.mfInfo.Quota = quotaRes
+
+			*req = *(req.WithContext(context.WithValue(req.Context(), util.MFInfokKey, mfInfo)))
 		}
+
 	}
 
 	rqh.next.ServeHTTP(rw, req)
 }
+
+// tryLockTag locks tag with redis ...
+func (rqh *regQuotaHandler) tryLockTag() (*common_redis.Mutex, error) {
+	con, err := redis.DialURL(
+		config.GetRedisOfRegURL(),
+		redis.DialConnectTimeout(dialConnectionTimeout),
+		redis.DialReadTimeout(dialReadTimeout),
+		redis.DialWriteTimeout(dialWriteTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+	tagLock := common_redis.New(con, rqh.mfInfo.Repository+":"+rqh.mfInfo.Tag, common_util.GenerateRandomString())
+	success, err := tagLock.Require()
+	if err != nil {
+		return nil, err
+	}
+	if !success {
+		return nil, fmt.Errorf("unable to lock tag: %s ", rqh.mfInfo.Repository+":"+rqh.mfInfo.Tag)
+	}
+	return tagLock, nil
+}
+
+func (rqh *regQuotaHandler) tryFreeTag() {
+	_, err := rqh.mfInfo.TagLock.Free()
+	if err != nil {
+		log.Warningf("Error to unlock tag: %s, with error: %v ", rqh.mfInfo.Tag, err)
+	}
+}
+
+// check the existence of a artifact, if exist, the method will return the artifact model
+func (rqh *regQuotaHandler) imageExist() (exist bool, af *models.Artifact, err error) {
+	artifactQuery := &models.ArtifactQuery{
+		PID:  rqh.mfInfo.ProjectID,
+		Repo: rqh.mfInfo.Repository,
+		Tag:  rqh.mfInfo.Tag,
+	}
+	afs, err := dao.ListArtifacts(artifactQuery)
+	if err != nil {
+		log.Errorf("Error occurred when to get project ID %v", err)
+		return false, nil, err
+	}
+	if len(afs) > 0 {
+		return true, afs[0], nil
+	}
+	return false, nil, nil
+}
+
+func (rqh *regQuotaHandler) tryRequireQuota() (quota.ResourceList, error) {
+	var quotaRes quota.ResourceList
+	quotaRes = quota.ResourceList{
+		quota.ResourceStorage: 0,
+		quota.ResourceCount:   0,
+	}
+
+	if !rqh.mfInfo.Exist {
+		// For manifest, we need to request 1 count and size for storage.
+		quotaRes = quota.ResourceList{
+			quota.ResourceStorage: rqh.mfInfo.Size,
+			quota.ResourceCount:   1,
+		}
+	} else if rqh.mfInfo.DigestChanged {
+		quotaRes = quota.ResourceList{
+			quota.ResourceStorage: rqh.mfInfo.Size,
+			quota.ResourceCount:   0,
+		}
+	}
```

Comment:
For the other cases(not included in the `if else block`), we shouldn't try to change the quota at all.